File: 491. Non-decreasing Subsequences/backtracking.cpp
class Solution {
public:
    void func(int index, vector<int>& nums, vector<int>& a, set<vector<int>>& ans) {
        int n = nums.size();
        
        if (index >= n) {
            if (a.size() >= 2)
                ans.insert(a);
            return;
        }
        
        if (!a.size() || nums[index] >= a.back()) {
            a.push_back(nums[index]);
            func(index + 1, nums, a, ans);
            a.pop_back();
        }
    
        func(index + 1, nums, a, ans);
    }
    
    vector<vector<int>> findSubsequences(vector<int>& nums) {
        vector<int> temp;
        set<vector<int>> ans;
        func(0, nums, temp, ans);
        return vector(ans.begin(), ans.end());
    }
    
};

File: 491. Non-decreasing Subsequences/backtracking.py
class Solution:
    def findSubsequences(self, nums: List[int]) -> List[List[int]]:
     
        '''
        In Python 2.7, we need to change the following line to:
        
        global res
        
        res = set()
        '''
        res = set()

        def backtracking(index, subsequence):
            '''
            In Python 2.7, we need to change the following line to:

            global res
            '''
            nonlocal res

            if len(subsequence) > 1:
                res.add(tuple(subsequence))

            if index == len(nums):
                return

            if not subsequence or nums[index] >= subsequence[-1]:
                backtracking(index + 1, subsequence + [nums[index]])

            backtracking(index + 1, subsequence)

        backtracking(0, [])

        return res